app/main.py

```diff
@@ -1,34 +1,116 @@
+from typing import List
+
+
 class Deck:
-    def __init__(self, row, column, is_alive=True):
-        pass
+    def __init__(self, row: int, column: int, is_alive: bool = True) -> None:
+        self.row = row
+        self.column = column
+        self.is_alive = is_alive
+
+    def __repr__(self) -> str:
+        return f"({self.row}, {self.column})"
 
 
 class Ship:
-    def __init__(self, start, end, is_drowned=False):
-        # Create decks and save them to a list `self.decks`
-        pass
+    def __init__(
+            self,
+            start: tuple,
+            end: tuple,
+            is_drowned: bool = False
+    ) -> None:
+        self.decks = self.set_up(start, end)
+        self.is_drowned = is_drowned
+
+    def get_deck(self, row: int, column: int) -> Deck:
+        for deck in self.decks:
+            if deck.row == row and deck.column == column:
+                return deck
 
-    def get_deck(self, row, column):
-        # Find the corresponding deck in the list
-        pass
+    @staticmethod
+    def set_up(start: tuple, end: tuple) -> List[Deck]:
+        length = (end[0] - start[0] + end[1] - start[1]) + 1
+        decks = []
+        for i in range(length):
+            if end[0] > end[1]:
+                decks.append(Deck(start[0] + i, start[1]))
+            else:
+                decks.append(Deck(start[0], start[1] + i))
+        return decks
 
-    def fire(self, row, column):
-        # Change the `is_alive` status of the deck
-        # And update the `is_drowned` value if it's needed
-        pass
+    def fire(self, row: int, column: int) -> None:
+        fire = self.get_deck(row, column)
+        if fire:
+            fire.is_alive = False
+            i = 0
+            for deck in self.decks:
+                if deck.is_alive is True:
+                    i += 1
+            if i == 0:
+                self.is_drowned = True
 
 
 class Battleship:
-    def __init__(self, ships):
-        # Create a dict `self.field`.
-        # Its keys are tuples - the coordinates of the non-empty cells,
-        # A value for each cell is a reference to the ship
-        # which is located in it
-        pass
-
-    def fire(self, location: tuple):
-        # This function should check whether the location
-        # is a key in the `self.field`
-        # If it is, then it should check if this cell is the last alive
-        # in the ship or not.
-        pass
+    def __init__(self, ships: list) -> None:
+        field = {}
+        for coord in ships:
+            ship = Ship(*coord)
+            for deck in ship.decks:
+                field[(deck.row, deck.column)] = ship
```

Comment:
i guess it would be better to remove this logic out of __init__ . something like your set_up method from the Ship class is nicer imo. Also that way you could've done your validate_field checks after setting things up (if you need it done automatically of course)

Reply:
You are absolutly right